src/components/ui/InfiniteList/index.tsx

```diff
@@ -0,0 +1,90 @@
+
+export interface InfiniteListProps extends Omit<HTMLAttributes<HTMLUListElement>, 'css'>, BaseProps {
+    /* The elements that will populat the list. These elements will be wrapped in <li> tags, so that can be any element. */
+    items: ReactNode[];
+    /* A callback function that will make an api call to load the next batch of items. */
+    handleLoad: () => void;
+    /* The number of items that will be loaded on each load. */
+    batchSize: number;
+}
+
+const InfiniteList = (props: InfiniteListProps) => {
+    const { items, handleLoad, batchSize } = props;
+
+    const feedRef = useRef<HTMLUListElement>(null);
+    const listEnd = useRef<HTMLDivElement>(null);
+    const currentTopItem = useRef<HTMLLIElement>(null);
+    const firstNew = useRef<HTMLLIElement>(null);
+    const prevLength = useRef(items.length)
+    const newLength = useRef(0);
+
+    const [isLoading, setIsLoading] = useState(false);
+
+    // Scroll to the bottom of the list on load
+    useEffect(() => {
+        feedRef.current?.addEventListener('scroll', handleScroll);
+        scrollToBottom();
+        return feedRef.current?.removeEventListener('scroll', handleScroll, true)
+    }, []);
+
+    const handleScroll = () => {
+        // when the scroll reaches the top
+        if (feedRef.current?.scrollTop === 0) {
+            setIsLoading(true)
+            setTimeout(
+                () => {
+                    handleLoad()
+                    setIsLoading(false)
+                    // scroll back down to the first new message
+                    firstNew.current?.scrollIntoView();
+                },
+                // simulate an async messages fetch
+                500
+            );
+        }
+    };
+
+    const scrollToBottom = () => {
+        listEnd.current?.scrollIntoView()
+    };
+
+    useEffect(() => {
+        if (items.length !== prevLength.current) {
+            prevLength.current = newLength.current;
+        }
+        newLength.current = items.length;
+    });
+
+    return (
+        <StyledInfiniteList ref={feedRef} className={isLoading ? 'ch-not-scrollable' : ''} {...props} data-testid='infinite-list'>
+            {isLoading && <li className="spinner"><Spinner /></li>}
+            {items.map((item, index) => {
+                if (index === 0) {
+                    return <li key={index} ref={currentTopItem}>{item}</li>
+                }
+                if (
+                    // It's not the initial load
+                    items.length > batchSize &&
+                    // It's the first message from the new batch
+                    index === newLength.current - prevLength.current - 1 &&
+                    // New messages were loaded
+                    newLength.current !== prevLength.current
+                ) {
+                    return <li key={index} ref={firstNew}>{item}</li>
+                }
+                return <li key={index}>{item}</li>
```

Comment:
It might be cleaner to have a method that decides what the `ref` value is for the `<li>` and assign `null` when it doesn't match the other 2 cases. So:

```
  getRef = () => {
    // all of this logic is now outside of the render block
  }
 <li key={index} ref={getRef()}>{item{/li>
```

Reply:
Do we actually need `firstNew`? It looks like we're just using it to scroll back into view, but what's the reason we can't just scroll to the `listEnd` element below and get rid of all of this logic?

Reply:
`firstNew` ensures that the first item in the new batch is scrolled into view. In my story included here, you have 250 items in a list. First you load with item 201-250 showing. Once you scroll up and the next batch is loaded, item 200 will be at the top of the visible list.